README.md
# TunSafe
Source code of the TunSafe client.

This open sourced TunSafe code is AGPL-1.0 licensed. Do note that the repository contains BSD and OpenSSL licensed files, so if you want to release a version based off of this repository you need to take that into account.

To build on Windows, open TunSafe.sln and build, or run build.py.

To build on Linux, run build_linux.sh

To build on FreeBSD, run build_freebsd.sh

## Docker
To use in docker, Just create you configuration file (for example: my-wg1.conf), then
```sh
docker run \
    -it \
    --rm \
    -v `pwd`/my-wg1.conf:/tmp/my-wg1.conf \
    --device /dev/net/tun \
    --cap-add NET_ADMIN \
    --network host \
    moghaddas/tunsafe start /tmp/my-wg1.conf
```

To Build:
```sh
git clone https://github.com/sinamoghaddas/TunSafe
cd TunSafe

docker build -t tunsafe .
```


installer/ChangeLog.txt
2018-12-16 - TunSafe v1.5-rc2

Changes:
1.Don't add endpoint route if route is not in included_routes
2.In BSD network code, don't add a route that's a subset of an Address
3.Don't add Excluded routes when Table=off
4.Display packet loss in Windows UI
5.Enable DNS block only if the DNS addr is a part of the routes
6.Support for WireGuard over TCP. Use Endpoint=tcp:// to connect
  to a TCP server, and use ListenPortTCP=12345 to listen on TCP.
7.Add support for Two Factor authentication. Read more on:
  https://github.com/TunSafe/TunSafe/wiki/Two-Factor-Authentication-with-TunSafe
8.Add support for a hybrid TCP/UDP mode that uses TCP for handshakes
  and UDP for data traffic. This means that PersistentKeepalive can
  be significantly raised to for example 300 seconds, since as long
  as the TCP connection remains open through NAT then the WireGuard
  connection will stay alive. Enable with Features=hybrid_tcp
9.Support for obfuscated WireGuard connections. Use ObfuscateKey=foo
  in the [Interface] section to setup the obfuscator key. It needs
  to be set to the same thing on both sides. There's also another
  setting to masquerade TCP connections as TLS. Use ObfuscateTCP to
  setup how TCP gets obfuscated. The default is to just make everything
  look totally random. It can also be set to tls-chrome or tls-firefox
  to make the traffic look like HTTPS traffic.
10.Display incoming invalid packets in Windows UI

2018-10-21 - TunSafe v1.5-rc1

Changes:
1.The kill switch is now remembered across computer restarts and
  is deactivated when disconnecting. Without this behavior, the 
  kill switch is unusable when auto connecting on Windows startup.
2.The kill switch is now optionally turned off on disconnect,
  and a button is shown in the UI to turn off the kill switch.
3.The kill switch can be configured to don't block local networks,
  this is used only for the firewall based kill switch.
4.Allow multiple DNS servers
5.Now the 'tunsafe' command line tool exists, which supports
  wg compatible configuration and statistics printing. The names
  of the adapters are the same as the adapters in the Control
  Panel network settings. It's used only when TunSafe runs
  in service mode.
6.The 'tunsafe' command line tool supports multiple wireguard
  sessions simultaneously using different tun interfaces.
7.Optimize IpToPeerMap for faster lookup using a trie.
8.Print a notice if a route we're trying to add already exists,
  perhaps will make it easier to debug issues.
9.Resolve DNS queries using a background thread, to make it
  possible to interrupt slow DNS queries.
10.IPv6 endpoint was printed incorrectly on the Advanced tab
11.Show an error message and drop packets if the TUN queue grows
   too large. This is a problem with the TAP NDIS6 driver on Win7.
12.Bundle the TunSafe-TAP installer instead of downloading it.
13.Don't show empty directories in the server list.

2018-10-08 - TunSafe v1.4

Changes:
1.Show a couple of more error strings when failing to edit registry.
2.Allow # comments in config file inside of a line
3.Enable persistent keepalive timer on sent handshakes

2018-08-11 - TunSafe v1.4-rc1

Changes:
1.Subfolders in the Config/ directory now show up as submenus.
2.Added a way to run TunSafe as a Windows Service.
  Foreground Mode: The service will disconnect when TunSafe closes.
  Background Mode: The service will stay connected in the background.
  No longer required to run the TunSafe client as Admin as long as
  the service is running.
3.New config setting [Interface].ExcludedIPs to configure IPs that
  should not be routed through TunSafe.
4.Can now automatically start TunSafe when Windows starts
5.New UI with tabs and graphs
6.Cache DNS queries to ensure DNS will succeed if connection fails
7.Recreate tray icon when explorer.exe restarts
8.Renamed window title to TunSafe instead of TunSafe VPN Client
9.Main window is now resizable
10.Disallow roaming endpoint when using AllowedIPs=0.0.0.0/0
   Only the original endpoint is added in the routing table so
   this would result in an endless loop of packets.
11.Display approximate Wireguard framing overhead in stats
12.Preparations for protocol handling with multiple threads
13.Delete the routes we made when disconnecting
14.Fix error message about unable to delete a route when connecting

2018-06-20 - TunSafe v1.3-rc3

Changes:
1.Add option to block Internet traffic outside of TunSafe. Either
  based on firewall rules, or by adding a null route, or both.
  The firewall rule blocks all traffic except traffic from TunSafe,
  loopback traffic, and DHCP traffic on the default NIC.
  The route rule adds two /1 routes to 0.0.0.0.
2.Convert LF to CRLF when importing config files
3.Update some logging messages
4.Delete the old routing rule pointing at the VPN server IP when
  disconnecting
5.Delete any conflicting old routing rule pointing at the VPN server
  when connecting.
6.Tray popup menu did not disappear when clicking outside of it.
7.Show config file names also in tray popup menu.
8.Make the menu item bold if connection is selected in popup menu.
9.Don't show the .conf filename extension in the UI.
10.Show also config file name when hovering on tray icon.
11.Click on the connected server to toggle connection
12.Fix bug where internet blocking checkbox was not removed.
13.Change so bold is used for selected server, and checkbox
   is used when connected.
14.Use WS_EX_COMPOSITED to reduce flicker
15.Now possible to enter a filename on command line to connect to.
16.Support /minimize and /minimize_on_connect command line opts.
17.Support PreUp,PostUp,PreDown,PostDown options on [Interface]
   Note: For security reasons you need to first enable them,
   so either Shift-Click on Options and select Allow Pre/Post Commands
   or specify the /allow_pre_post command line option.

2018-04-29 - TunSafe v1.2

Changes:
1.Use /24 instead of failing when a /32 Address is used
2.Use /120 instead of failing when a /128 Address is used

util.cpp
// SPDX-License-Identifier: AGPL-1.0-only
// Copyright (C) 2018 Ludvig Strigeus <info@tunsafe.com>. All Rights Reserved.
#include "stdafx.h"

#include <assert.h>
#include <stdlib.h>
#include <stdarg.h>
#include <string.h>
#include <string>

#if defined(OS_POSIX)
#include <unistd.h>
#include <fcntl.h>
#include <sys/types.h>
#include <sys/wait.h>
#include <netinet/in.h>
#include <arpa/inet.h>
#include <sys/time.h>
#endif


#if defined(OS_MACOSX)
#include <mach/mach_time.h>
#endif  // OS_MACOSX

#include <vector>
#include <algorithm>
#include "tunsafe_types.h"
#include "tunsafe_endian.h"

static const char kBase64Alphabet[] = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/";

char *base64_encode(const uint8 *input, size_t length, char *output, size_t output_size, size_t *out_length) {
  char *result, *r;
  const uint8 *end;

  size_t size = (length + 2) / 3 * 4 + 1;

  if (output != NULL) {
    result = output;
    assert(output_size >= size);
    if (output_size < size) {
      *result = 0;
      return NULL;
    }
  } else {
    result = (char*)malloc(size);
    if (!result)
      return NULL;
  }
  r = result;
  end = input + length - 3;
  // Encode full blocks
  while (input <= end) {
    uint32 a = (input[0] << 16) + (input[1] << 8) + input[2];
    input += 3;

    r[0] = kBase64Alphabet[(a >> 18)/* & 0x3F*/];
    r[1] = kBase64Alphabet[(a >> 12) & 0x3F];
    r[2] = kBase64Alphabet[(a >> 6) & 0x3F];
    r[3] = kBase64Alphabet[(a) & 0x3F];
    r += 4;
  }
  if (input == end + 2) {
    uint32 a = input[0] << 4;
    r[0] = kBase64Alphabet[(a >> 6) /*& 0x3F*/];
    r[1] = kBase64Alphabet[(a) & 0x3F];
    r[2] = '=';
    r[3] = '=';
    r += 4;
  } else if (input == end + 1) {
    uint32 a = (input[0] << 10) + (input[1] << 2);
    r[0] = kBase64Alphabet[(a >> 12) /*& 0x3F*/];
    r[1] = kBase64Alphabet[(a >> 6) & 0x3F];
    r[2] = kBase64Alphabet[(a) & 0x3F];
    r[3] = '=';
    r += 4;
  }
  if (out_length)
    *out_length = r - result;
  *r = 0;
  return result;
}

#define WHITESPACE 64
#define EQUALS     65
#define INVALID    66

static const unsigned char d[] = {
  66,66,66,66,66,66,66,66,66,66,64,66,66,66,66,66,66,66,66,66,66,66,66,66,66,
  66,66,66,66,66,66,66,66,66,66,66,66,66,66,66,66,66,66,62,66,66,66,63,52,53,
  54,55,56,57,58,59,60,61,66,66,66,65,66,66,66, 0, 1, 2, 3, 4, 5, 6, 7, 8, 9,
  10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25,66,66,66,66,66,66,26,27,28,
  29,30,31,32,33,34,35,36,37,38,39,40,41,42,43,44,45,46,47,48,49,50,51,66,66,
  66,66,66,66,66,66,66,66,66,66,66,66,66,66,66,66,66,66,66,66,66,66,66,66,66,
  66,66,66,66,66,66,66,66,66,66,66,66,66,66,66,66,66,66,66,66,66,66,66,66,66,
  66,66,66,66,66,66,66,66,66,66,66,66,66,66,66,66,66,66,66,66,66,66,66,66,66,
  66,66,66,66,66,66,66,66,66,66,66,66,66,66,66,66,66,66,66,66,66,66,66,66,66,
  66,66,66,66,66,66,66,66,66,66,66,66,66,66,66,66,66,66,66,66,66,66,66,66,66,
  66,66,66,66,66,66
};

bool base64_decode(uint8 *in, size_t inLen, uint8 *out, size_t *outLen) {
  uint8 *end = in + inLen;
  uint8 iter = 0;
  uint32_t buf = 0;
  size_t len = 0;

  while (in < end) {
    unsigned char c = d[*in++];

    switch (c) {
    case WHITESPACE: continue;   /* skip whitespace */
    case INVALID:    return false;   /* invalid input, return error */
    case EQUALS:                 /* pad character, end of data */
      in = end;
      continue;
    default:
      buf = buf << 6 | c;
      iter++;
      if (iter == 4) {
        if ((len += 3) > *outLen) return 0; /* buffer overflow */
        *(out++) = (buf >> 16) & 255;
        *(out++) = (buf >> 8) & 255;
        *(out++) = buf & 255;
        buf = 0; iter = 0;

      }
    }
  }
  if (iter == 3) {
    if ((len += 2) > *outLen) return 0; /* buffer overflow */
    *(out++) = (buf >> 10) & 255;
    *(out++) = (buf >> 2) & 255;
  } else if (iter == 2) {
    if (++len > *outLen) return 0; /* buffer overflow */
    *(out++) = (buf >> 4) & 255;
  }
  *outLen = len;
  return true;
}



int RunCommand(const char *fmt, ...) {
  const char *fmt_org = fmt;
  va_list va;
  std::string tmp;
  char buf[32], c;
  char *args[33];
  char *envp[1] = {NULL};
  int nargs = 0;
  bool didadd = false;
  va_start(va, fmt);
  for (;;) {
    c = *fmt++;
    if (c == '%') {
      c = *fmt++;
      if (c == 0) goto ZERO;
      if (c == 's') {
        char *arg = va_arg(va, char*);
        if (arg != NULL) {
          tmp += arg;
          didadd = true;
        }
      } else if (c == 'd') {
        snprintf(buf, 32, "%d", va_arg(va, int));
        tmp += buf;
      } else if (c == '%') {
        tmp += '%';
      } else if (c == 'A') {
        struct in_addr in;
        in.s_addr = htonl(va_arg(va, in_addr_t));
        tmp += inet_ntoa(in);
      }
    } else if (c == ' ' || c == 0) {
ZERO:
      if (!tmp.empty() || didadd) {
        args[nargs++] = _strdup(tmp.c_str());
        tmp.clear();
        if (nargs == 32 || c == 0) break;
      }
      didadd = false;
    } else {
      tmp += c;
    }
  }
  args[nargs] = 0;

  fprintf(stderr, "Run:");
  for (int i = 0; args[i]; i++)
    fprintf(stderr, " %s", args[i]);
  fprintf(stderr, "\n");

  int ret = -1;


#if defined(OS_POSIX)
  pid_t pid = fork();
  if (pid == 0) {
    execve(args[0], args, envp);
    exit(127);
  }
  if (pid < 0) {
    RERROR("Fork failed");
  } else if (waitpid(pid, &ret, 0) != pid) {
    ret = -1;
  }
#endif

  if (ret != 0)
    RERROR("Command failed %d!", ret);

  return ret;
}

bool IsOnlyZeros(const uint8 *data, size_t data_size) {
  for (size_t i = 0; i != data_size; i++)
    if (data[i])
      return false;
  return true;
}


#ifdef _MSC_VER
void printhex(const char *name, const void *a, size_t l) {
  char buf[256];
  snprintf(buf, 256, "%s (%d):", name, (int)l); OutputDebugString(buf);
  for (size_t i = 0; i < l; i++) {
    if (i % 4 == 0) printf(" ");
    snprintf(buf, 256, "%.2X", *((uint8*)a + i)); OutputDebugString(buf);
  }
  OutputDebugString("\n");
}

#else
void printhex(const char *name, const void *a, size_t l) {
  printf("%s (%d):", name, (int)l);
  for (size_t i = 0; i < l; i++) {
    if (i % 4 == 0) printf(" ");
    printf("%.2X", *((uint8*)a + i));
  }
  printf("\n");
}
#endif

typedef void Logger(int type, const char *msg);
Logger *g_logger;

#undef RERROR
#undef void 

void RERROR(const char *msg, ...);

void RERROR(const char *msg, ...) {
  va_list va;
  char buf[512];
  va_start(va, msg);
  vsnprintf(buf, sizeof(buf), msg, va);
  va_end(va);
  if (g_logger) {
    g_logger(1, buf);
  } else {
    fputs(buf, stderr);
    fputs("\n", stderr);
  }
}


void RINFO(const char *msg, ...) {
  va_list va;
  char buf[512];
  va_start(va, msg);
  vsnprintf(buf, sizeof(buf), msg, va);
  va_end(va);
  if (g_logger) {
    g_logger(0, buf);
  } else {
    fputs(buf, stderr);
    fputs("\n", stderr);
  }
}

void *memdup(const void *p, size_t size) {
  void *x = malloc(size);
  if (x)
    memcpy(x, p, size);
  return x;
}

char *my_strndup(const char *p, size_t size) {
  char *x = (char*)malloc(size + 1);
  if (x) {
    x[size] = 0;
    memcpy(x, p, size);
  }
  return x;
}

size_t my_strlcpy(char *dst, size_t dstsize, const char *src) {
  size_t len = strlen(src);
  if (dstsize) {
    size_t lenx = std::min<size_t>(dstsize - 1, len);
    dst[lenx] = 0;
    memcpy(dst, src, lenx);
  }
  return len;
}

void OsGetRandomBytes(uint8 *data, size_t data_size) {
#if defined(OS_WIN)
  static BOOLEAN(APIENTRY *pfn)(void*, ULONG);
  if (!pfn) {
    pfn = (BOOLEAN(APIENTRY *)(void*, ULONG))GetProcAddress(LoadLibrary("ADVAPI32.DLL"), "SystemFunction036");
    if (!pfn)
      ExitProcess(1);
  }
  if (!pfn(data, (ULONG)data_size)) {
    ExitProcess(1);
    return;
  }
#elif defined(OS_POSIX)
  int fd = open("/dev/urandom", O_RDONLY);
  if (fd < 0) {
    fprintf(stderr, "/dev/urandom failed\n");
    exit(1);
  }
  int r = read(fd, data, data_size);
  if (r != data_size) {
    fprintf(stderr, "/dev/urandom failed\n");
    exit(1);
  }
  close(fd);
#else
#error
#endif
}

bool ParseConfigKeyValue(char *m, std::vector<std::pair<char *, char*>> *result) {
  for (;;) {
    char *nl = strchr(m, '\n');
    if (nl)
      *nl = 0;
    if (*m != '\0') {
      char *value = strchr(m, '=');
      if (value == NULL)
        return false;
      *value++ = '\0';
      result->emplace_back(m, value);
    }
    if (!nl)
      return true;
    m = nl + 1;
  }
}

bool ParseHexString(const char *text, void *data, size_t data_size) {
  size_t len = strlen(text);
  if (len != data_size * 2)
    return false;
  for (size_t i = 0; i < data_size; i++) {
    uint32 c = text[i * 2 + 0];
    if (c >= '0' && c <= '9') {
      c -= '0';
    } else if ((c |= 32) >= 'a' && c <= 'f') {
      c -= 'a' - 10;
    } else {
      return false;
    }
    uint32 d = text[i * 2 + 1];
    if (d >= '0' && d <= '9') {
      d -= '0';
    } else if ((d |= 32) >= 'a' && d <= 'f') {
      d -= 'a' - 10;
    } else {
      return false;
    }
    ((uint8*)data)[i] = c * 16 + d;
  }
  return true;
}

bool is_space(uint8_t c) {
  return c == ' ' || c == '\r' || c == '\n' || c == '\t';
}

void SplitString(char *s, int separator, std::vector<char*> *components) {
  components->clear();
  for (;;) {
    while (is_space(*s)) s++;
    char *d = strchr(s, separator);
    if (d == NULL) {
      if (*s)
        components->push_back(s);
      return;
    }
    *d = 0;
    char *e = d;
    while (e > s && is_space(e[-1]))
      *--e = 0;
    components->push_back(s);
    s = d + 1;
  }
}

void PrintHexString(const void *data, size_t data_size, char *result) {
  for (size_t i = 0; i < data_size; i++) {
    uint8 c = ((uint8*)data)[i];
    *result++ = "0123456789abcdef"[c >> 4];
    *result++ = "0123456789abcdef"[c & 0xF];
  }
  *result++ = 0;
}

bool ParseBase64Key(const char *s, uint8 key[32]) {
  size_t size = 32;
  return base64_decode((uint8*)s, strlen(s), key, &size) && size == 32;
}


#if defined(OS_WIN)
uint64 OsGetMilliseconds() {
  return GetTickCount64();
}

void OsGetTimestampTAI64N(uint8 dst[12]) {
  SYSTEMTIME systime;
  uint64 file_time_uint64 = 0;
  GetSystemTime(&systime);
  SystemTimeToFileTime(&systime, (FILETIME*)&file_time_uint64);
  uint64 time_since_epoch_100ns = (file_time_uint64 - 116444736000000000);
  uint64 secs_since_epoch = time_since_epoch_100ns / 10000000 + 0x400000000000000a;
  uint32 nanos = (uint32)(time_since_epoch_100ns % 10000000) * 100;
  WriteBE64(dst, secs_since_epoch);
  WriteBE32(dst + 8, nanos);
}

void OsInterruptibleSleep(int millis) {
  SleepEx(millis, TRUE);
}

#endif  // defined(OS_WIN)


#if defined(OS_POSIX)

#if defined(OS_MACOSX)
static mach_timebase_info_data_t timebase = { 0, 0 };
static uint64_t                  initclock;

void InitOsxGetMilliseconds() {
  if (mach_timebase_info(&timebase) != 0)
    abort();
  initclock = mach_absolute_time();

  timebase.denom *= 1000000;
}

uint64 OsGetMilliseconds() {
  assert(initclock != 0);

  uint64_t clock = mach_absolute_time() - initclock;
  return clock * (uint64_t)timebase.numer / (uint64_t)timebase.denom;
}

#else  // defined(OS_MACOSX)
uint64 OsGetMilliseconds() {
  struct timespec ts;
  if (clock_gettime(CLOCK_MONOTONIC, &ts) != 0) {
    //error
    fprintf(stderr, "clock_gettime failed\n");
    exit(1);
  }
  return (uint64)ts.tv_sec * 1000 + (ts.tv_nsec / 1000000);
}
#endif

void OsGetTimestampTAI64N(uint8 dst[12]) {
  struct timeval tv;
  gettimeofday(&tv, NULL);
  uint64 secs_since_epoch = tv.tv_sec + 0x400000000000000a;
  uint32 nanos = tv.tv_usec * 1000;
  WriteBE64(dst, secs_since_epoch);
  WriteBE32(dst + 8, nanos);
}

void OsInterruptibleSleep(int millis) {
  usleep((useconds_t)millis * 1000);
 }

#endif  // defined(OS_POSIX)
